File: src/python/xerxes/runtime/cost_tracker.py
# Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""In-session LLM cost accounting.

Provides :class:`CostEvent` (one billing record) and :class:`CostTracker`
(an append-only ledger held by every :class:`QueryEngine`). Costs are
computed via :func:`xerxes.llms.registry.calc_cost`; prompt-caching reads
and writes are recorded so the ``/cost`` and ``/usage`` slash commands can
report cache-hit rates and discounted spend.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime
from typing import Any


@dataclass
class CostEvent:
    """One immutable cost ledger entry.

    Attributes:
        model: LLM identifier the event was billed against.
        in_tokens: Non-cached input tokens.
        out_tokens: Output tokens.
        cost_usd: Dollar cost computed at record time.
        label: Free-form label, typically ``"turn_<n>"``.
        timestamp: ISO-8601 timestamp captured at construction.
        cache_read_tokens: Tokens served from a prompt-cache hit.
        cache_creation_tokens: Tokens written into a new cache entry.
    """

    model: str
    in_tokens: int
    out_tokens: int
    cost_usd: float
    label: str = ""
    timestamp: str = field(default_factory=lambda: datetime.now().isoformat())
    cache_read_tokens: int = 0
    cache_creation_tokens: int = 0


@dataclass
class CostTracker:
    """Append-only ledger of :class:`CostEvent` entries for one session.

    Attributes:
        events: All cost events recorded so far, in insertion order.
    """

    events: list[CostEvent] = field(default_factory=list)

    def record_turn(
        self,
        model: str,
        in_tokens: int,
        out_tokens: int,
        label: str = "",
        *,
        cache_read_tokens: int = 0,
        cache_creation_tokens: int = 0,
    ) -> CostEvent:
        """Record a single LLM turn's token usage and append a priced event.

        Args:
            model: LLM identifier used to look up per-token pricing.
            in_tokens: Non-cached input tokens consumed by the turn.
            out_tokens: Output tokens produced by the turn.
            label: Optional human-friendly label stored on the event.
            cache_read_tokens: Tokens served from a prompt cache; billed at
                ~10% of the normal input rate.
            cache_creation_tokens: Tokens used to populate a new cache entry;
                billed at ~125% of the normal input rate.

        Returns:
            The newly appended :class:`CostEvent`.

        Note:
            ``calc_cost`` does not yet model cache pricing natively; raw
            cache counts are still recorded so the pricing fix can be
            replayed retroactively.
        """

        from xerxes.llms.registry import calc_cost

        # Approximate cache-aware cost: pretend cache_read is 10% of
        # normal input tokens, and cache_creation is 125%.
        effective_in = in_tokens
        cache_extra_cost = 0.0
        if cache_read_tokens or cache_creation_tokens:
            per_in_unit_cost = (calc_cost(model, 1000, 0) / 1000.0) if calc_cost(model, 1000, 0) > 0 else 0.0
            cache_extra_cost = (cache_read_tokens * per_in_unit_cost * 0.1) + (
                cache_creation_tokens * per_in_unit_cost * 1.25
            )
        cost = calc_cost(model, effective_in, out_tokens) + cache_extra_cost
        event = CostEvent(
            model=model,
            in_tokens=in_tokens,
            out_tokens=out_tokens,
            cost_usd=cost,
            label=label,
            cache_read_tokens=cache_read_tokens,
            cache_creation_tokens=cache_creation_tokens,
        )
        self.events.append(event)
        return event

    def record_raw(self, label: str, cost_usd: float, model: str = "") -> CostEvent:
        """Record an externally-priced cost (e.g. embeddings, image gen).

        Use this when the caller already knows the dollar amount and has no
        meaningful token counts to attribute.
        """

        event = CostEvent(
            model=model,
            in_tokens=0,
            out_tokens=0,
            cost_usd=cost_usd,
            label=label,
        )
        self.events.append(event)
        return event

    @property
    def total_cost_usd(self) -> float:
        """Cumulative dollar cost summed across every recorded event."""
        return sum(e.cost_usd for e in self.events)

    @property
    def total_input_tokens(self) -> int:
        """Total non-cached input tokens consumed across every event."""
        return sum(e.in_tokens for e in self.events)

    @property
    def total_output_tokens(self) -> int:
        """Total output tokens produced across every event."""
        return sum(e.out_tokens for e in self.events)

    @property
    def total_tokens(self) -> int:
        """Sum of input and output tokens across every event."""
        return self.total_input_tokens + self.total_output_tokens

    @property
    def total_cache_read_tokens(self) -> int:
        """Total tokens served from a prompt-cache hit across every event."""
        return sum(e.cache_read_tokens for e in self.events)

    @property
    def total_cache_creation_tokens(self) -> int:
        """Total tokens written into new cache entries across every event."""
        return sum(e.cache_creation_tokens for e in self.events)

    def cache_hit_rate(self) -> float:
        """Return the fraction of input tokens served from the prompt cache.

        Returns ``0.0`` when no input tokens have been recorded yet.
        """
        served = self.total_cache_read_tokens + self.total_input_tokens
        if served <= 0:
            return 0.0
        return self.total_cache_read_tokens / float(served)

    @property
    def event_count(self) -> int:
        """Number of :class:`CostEvent` records held."""
        return len(self.events)

    def by_model(self) -> dict[str, dict[str, Any]]:
        """Return a per-model aggregate of tokens, cost, and turn count."""

        agg: dict[str, dict[str, Any]] = {}
        for e in self.events:
            if e.model not in agg:
                agg[e.model] = {"in_tokens": 0, "out_tokens": 0, "cost_usd": 0.0, "turns": 0}
            agg[e.model]["in_tokens"] += e.in_tokens
            agg[e.model]["out_tokens"] += e.out_tokens
            agg[e.model]["cost_usd"] += e.cost_usd
            agg[e.model]["turns"] += 1
        return agg

    def clear(self) -> None:
        """Drop every recorded :class:`CostEvent`."""
        self.events.clear()

    def summary(self) -> str:
        """Render a Markdown summary of totals plus per-model breakdown."""

        lines = [
            "# Cost Summary",
            "",
            f"Total cost: ${self.total_cost_usd:.4f}",
            f"Total tokens: {self.total_tokens:,} (in: {self.total_input_tokens:,}, out: {self.total_output_tokens:,})",
            f"Events: {self.event_count}",
            "",
        ]

        by_model = self.by_model()
        if by_model:
            lines.append("## By Model")
            for model, stats in sorted(by_model.items()):
                lines.append(
                    f"- **{model}**: ${stats['cost_usd']:.4f} "
                    f"({stats['turns']} turns, {stats['in_tokens'] + stats['out_tokens']:,} tokens)"
                )

        return "\n".join(lines)

    def as_dicts(self) -> list[dict[str, Any]]:
        """Serialise every event as a JSON-friendly dict for persistence."""

        return [
            {
                "model": e.model,
                "in_tokens": e.in_tokens,
                "out_tokens": e.out_tokens,
                "cost_usd": e.cost_usd,
                "label": e.label,
                "timestamp": e.timestamp,
            }
            for e in self.events
        ]


__all__ = [
    "CostEvent",
    "CostTracker",
]

File: xerxes/src/runtime/costTracker.ts
// Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
// Licensed under the Apache License, Version 2.0.

import type { QuerySource } from '../llms/client.js'
import { calcCost } from '../llms/providerRegistry.js'

const CACHE_READ_MULTIPLIER = 0.1
const CACHE_CREATION_MULTIPLIER = 1.25
const PRICING_PROBE_TOKENS = 1_000

/** Bucket used by aggregate views for events without an explicit scope. */
export const UNSCOPED_COST_SCOPE = '(unscoped)'

/**
 * The one source that is not housekeeping.
 *
 * Compared as a literal instead of importing the llms helper so the ledger
 * keeps a type-only dependency on the provider layer; the constant is typed as
 * {@link QuerySource}, so a typo or a renamed union member fails to compile.
 */
const MAIN_COST_SOURCE: QuerySource = 'main'

export type CostCalculator = (model: string, inputTokens: number, outputTokens: number) => number

export interface CostEventOptions {
  readonly agentId?: string
  readonly cacheCreationTokens?: number
  readonly cacheReadTokens?: number
  readonly costUsd: number
  readonly inputTokens: number
  readonly label?: string
  readonly model: string
  readonly outputTokens: number
  readonly sessionId?: string
  /** Why the call was made; absent on legacy events recorded before the dimension existed. */
  readonly source?: QuerySource
  readonly timestamp?: string
}

export interface CostEventRecord {
  readonly agent_id: string | null
  readonly cache_creation_tokens: number
  readonly cache_read_tokens: number
  readonly cost_usd: number
  readonly in_tokens: number
  readonly label: string
  readonly model: string
  readonly out_tokens: number
  readonly session_id: string | null
  readonly source: string | null
  readonly timestamp: string
}

export interface LegacyCostEventRecord {
  readonly cost_usd: number
  readonly in_tokens: number
  readonly label: string
  readonly model: string
  readonly out_tokens: number
  readonly timestamp: string
}

export interface RecordTurnOptions {
  readonly agentId?: string
  readonly cacheCreationTokens?: number
  readonly cacheReadTokens?: number
  readonly sessionId?: string
  readonly source?: QuerySource
  readonly timestamp?: string
}

export interface RecordRawOptions {
  readonly agentId?: string
  readonly sessionId?: string
  readonly source?: QuerySource
  readonly timestamp?: string
}

export interface CostTrackerOptions {
  /** Default identity applied to events that omit an agent ID. */
  readonly agentId?: string
  /** Provider/model cost function; defaults to the shared pricing registry. */
  readonly costCalculator?: CostCalculator
  /** Injectable wall-clock for deterministic timestamps. */
  readonly now?: () => Date
  /** Default identity applied to events that omit a session ID. */
  readonly sessionId?: string
  /** Default source applied to events that omit one, e.g. a compaction-owned tracker. */
  readonly source?: QuerySource
}

/** Filters narrowing a source breakdown to one session and/or agent. */
export interface CostSourceBreakdownOptions {
  readonly agentId?: string
  readonly sessionId?: string
}

/**
 * Answer to "how much of this spend was housekeeping".
 *
 * `untagged` is kept separate from both halves rather than folded into
 * housekeeping: events recorded before a call site was taught to pass a source
 * are unknown, not proven housekeeping, and merging them would overstate the
 * saving an auxiliary route appears to deliver.
 */
export interface CostSourceBreakdown {
  readonly bySource: Readonly<Record<string, CostAggregate>>
  readonly housekeeping: CostAggregate
  /** Housekeeping share of priced spend in [0, 1]; zero when nothing cost anything. */
  readonly housekeepingCostShare: number
  readonly main: CostAggregate
  readonly total: CostAggregate
  readonly untagged: CostAggregate
}

export interface CostAggregate {
  readonly cacheCreationTokens: number
  readonly cacheHitRate: number
  readonly cacheReadTokens: number
  readonly costUsd: number
  readonly inputTokens: number
  readonly outputTokens: number
  readonly tokens: number
  readonly turns: number
}

/**
 * Immutable record for one priced LLM or externally-priced operation.
 *
 * Token counts deliberately exclude cache reads from inputTokens, matching
 * the Python ledger. Cache reads and writes are retained separately so
 * pricing can be recalculated without losing raw provider usage.
 */
export class CostEvent {
  readonly agentId: string | undefined
  readonly cacheCreationTokens: number
  readonly cacheReadTokens: number
  readonly costUsd: number
  readonly inputTokens: number
  readonly label: string
  readonly model: string
  readonly outputTokens: number
  readonly sessionId: string | undefined
  readonly source: QuerySource | undefined
  readonly timestamp: string

  constructor(options: CostEventOptions) {
    this.model = stringValue(options.model, 'model')
    this.inputTokens = tokenCount(options.inputTokens, 'inputTokens')
    this.outputTokens = tokenCount(options.outputTokens, 'outputTokens')
    this.costUsd = finiteNumber(options.costUsd, 'costUsd')
    this.label = stringValue(options.label ?? '', 'label')
    this.timestamp = timestampValue(options.timestamp ?? new Date().toISOString())
    this.cacheReadTokens = tokenCount(options.cacheReadTokens ?? 0, 'cacheReadTokens')
    this.cacheCreationTokens = tokenCount(options.cacheCreationTokens ?? 0, 'cacheCreationTokens')
    this.sessionId = scopeValue(options.sessionId, 'sessionId')
    this.agentId = scopeValue(options.agentId, 'agentId')
    this.source = sourceValue(options.source)
    Object.freeze(this)
  }

  /** Full persistence form including cache usage and session/agent attribution. */
  toRecord(): CostEventRecord {
    return {
      model: this.model,
      in_tokens: this.inputTokens,
      out_tokens: this.outputTokens,
      cost_usd: this.costUsd,
      label: this.label,
      timestamp: this.timestamp,
      cache_read_tokens: this.cacheReadTokens,
      cache_creation_tokens: this.cacheCreationTokens,
      session_id: this.sessionId ?? null,
      agent_id: this.agentId ?? null,
      source: this.source ?? null,
    }
  }

  /** Python cost_tracker.py serialization shape, retained for old ledgers. */
  toLegacyRecord(): LegacyCostEventRecord {
    return {
      model: this.model,
      in_tokens: this.inputTokens,
      out_tokens: this.outputTokens,
      cost_usd: this.costUsd,
      label: this.label,
      timestamp: this.timestamp,
    }
  }
}

/**
 * Append-only LLM cost ledger with cache-aware pricing and scope aggregates.
 *
 * A tracker can serve one session by supplying sessionId at construction, or
 * a host-wide ledger by attaching sessionId and agentId to individual events.
 */
export class CostTracker {
  private readonly calculator: CostCalculator
  private readonly clock: () => Date
  private readonly defaultAgentId: string | undefined
  private readonly defaultSessionId: string | undefined
  private readonly defaultSource: QuerySource | undefined
  private readonly ledger: CostEvent[] = []

  constructor(options: CostTrackerOptions = {}) {
    this.calculator = options.costCalculator ?? calcCost
    this.clock = options.now ?? (() => new Date())
    this.defaultSessionId = scopeValue(options.sessionId, 'sessionId')
    this.defaultAgentId = scopeValue(options.agentId, 'agentId')
    this.defaultSource = sourceValue(options.source)
  }

  /** Snapshot event list; callers cannot mutate the underlying ledger array. */
  get events(): readonly CostEvent[] {
    return [...this.ledger]
  }

  get eventCount(): number {
    return this.ledger.length
  }

  get totalCostUsd(): number {
    return aggregate(this.ledger).costUsd
  }

  get totalInputTokens(): number {
    return aggregate(this.ledger).inputTokens
  }

  get totalOutputTokens(): number {
    return aggregate(this.ledger).outputTokens
  }

  get totalTokens(): number {
    return this.totalInputTokens + this.totalOutputTokens
  }

  get totalCacheReadTokens(): number {
    return aggregate(this.ledger).cacheReadTokens
  }

  get totalCacheCreationTokens(): number {
    return aggregate(this.ledger).cacheCreationTokens
  }

  /** Convenience alias for callers that only need total spend. */
  get totalCost(): number {
    return this.totalCostUsd
  }

  /**
   * Price and append one LLM completion event.
   *
   * Cache reads cost 10% of normal input-token price and cache creation costs
   * 125%, matching the Python implementation. The shared calcCost registry
   * keeps provider-prefixed model names and unknown-model zero-pricing intact.
   */
  recordTurn(
    model: string,
    inputTokens: number,
    outputTokens: number,
    label = '',
    options: RecordTurnOptions = {},
  ): CostEvent {
    const validatedModel = stringValue(model, 'model')
    const validatedInput = tokenCount(inputTokens, 'inputTokens')
    const validatedOutput = tokenCount(outputTokens, 'outputTokens')
    const cacheReadTokens = tokenCount(options.cacheReadTokens ?? 0, 'cacheReadTokens')
    const cacheCreationTokens = tokenCount(options.cacheCreationTokens ?? 0, 'cacheCreationTokens')
    const baseCost = finiteNumber(this.calculator(validatedModel, validatedInput, validatedOutput), 'calculated cost')
    let cacheCost = 0
    if (cacheReadTokens || cacheCreationTokens) {
      const inputProbe = finiteNumber(this.calculator(validatedModel, PRICING_PROBE_TOKENS, 0), 'input pricing')
      const inputRate = inputProbe > 0 ? inputProbe / PRICING_PROBE_TOKENS : 0
      cacheCost = cacheReadTokens * inputRate * CACHE_READ_MULTIPLIER
        + cacheCreationTokens * inputRate * CACHE_CREATION_MULTIPLIER
    }
    return this.append(new CostEvent({
      model: validatedModel,
      inputTokens: validatedInput,
      outputTokens: validatedOutput,
      costUsd: baseCost + cacheCost,
      label: stringValue(label, 'label'),
      timestamp: options.timestamp ?? this.nowTimestamp(),
      cacheReadTokens,
      cacheCreationTokens,
      ...(this.scopeOptions(options)),
    }))
  }

  /** Append a separately-priced operation such as embeddings or image generation. */
  recordRaw(label: string, costUsd: number, model = '', options: RecordRawOptions = {}): CostEvent {
    return this.append(new CostEvent({
      model: stringValue(model, 'model'),
      inputTokens: 0,
      outputTokens: 0,
      costUsd: finiteNumber(costUsd, 'costUsd'),
      label: stringValue(label, 'label'),
      timestamp: options.timestamp ?? this.nowTimestamp(),
      ...(this.scopeOptions(options)),
    }))
  }

  /** Append a pre-built immutable event, preserving its original timestamp and scopes. */
  record(event: CostEvent): CostEvent {
    if (!(event instanceof CostEvent)) throw new TypeError('event must be a CostEvent')
    return this.append(event)
  }

  /** Fraction of all served input tokens supplied by a prompt-cache hit. */
  cacheHitRate(): number {
    return cacheHitRate(this.ledger)
  }

  /** Aggregate events by billed model in insertion order. */
  byModel(): Readonly<Record<string, CostAggregate>> {
    return groupBy(this.ledger, event => event.model)
  }

  /** Aggregate events by session; events without a session use the exported unscoped bucket. */
  bySession(): Readonly<Record<string, CostAggregate>> {
    return groupBy(this.ledger, event => event.sessionId ?? UNSCOPED_COST_SCOPE)
  }

  /** Aggregate events by agent; events without an agent use the exported unscoped bucket. */
  byAgent(): Readonly<Record<string, CostAggregate>> {
    return groupBy(this.ledger, event => event.agentId ?? UNSCOPED_COST_SCOPE)
  }

  /** Aggregate events by call source; events recorded without one use the unscoped bucket. */
  bySource(): Readonly<Record<string, CostAggregate>> {
    return groupBy(this.ledger, event => event.source ?? UNSCOPED_COST_SCOPE)
  }

  /** Aggregate exactly one call source, returning zeroes when it has no events. */
  forSource(source: QuerySource): CostAggregate {
    return aggregate(this.ledger.filter(event => event.source === source))
  }

  /**
   * Split spend into main-loop, housekeeping, and untagged buckets.
   *
   * This is the accessor that makes an auxiliary route observable: without it
   * there is no way to tell whether compaction, titling, and memory extraction
   * moved off the main model or are still billing at main-model rates.
   * Housekeeping is defined as "tagged with any source other than main", so a
   * newly added source counts immediately instead of silently landing in the
   * main bucket until this file is updated.
   */
  sourceBreakdown(options: CostSourceBreakdownOptions = {}): CostSourceBreakdown {
    const sessionId = scopeValue(options.sessionId, 'sessionId')
    const agentId = scopeValue(options.agentId, 'agentId')
    const scoped = this.ledger.filter(event =>
      (sessionId === undefined || event.sessionId === sessionId)
      && (agentId === undefined || event.agentId === agentId))
    const total = aggregate(scoped)
    const housekeeping = aggregate(scoped.filter(event =>
      event.source !== undefined && event.source !== MAIN_COST_SOURCE))
    return Object.freeze({
      total,
      main: aggregate(scoped.filter(event => event.source === MAIN_COST_SOURCE)),
      housekeeping,
      untagged: aggregate(scoped.filter(event => event.source === undefined)),
      bySource: groupBy(scoped, event => event.source ?? UNSCOPED_COST_SCOPE),
      housekeepingCostShare: total.costUsd > 0 ? housekeeping.costUsd / total.costUsd : 0,
    })
  }

  /** Aggregate exactly one scoped session, returning zeroes when it has no events. */
  forSession(sessionId: string): CostAggregate {
    const expected = scopeValue(sessionId, 'sessionId')
    return aggregate(this.ledger.filter(event => event.sessionId === expected))
  }

  /** Aggregate exactly one scoped agent, returning zeroes when it has no events. */
  forAgent(agentId: string): CostAggregate {
    const expected = scopeValue(agentId, 'agentId')
    return aggregate(this.ledger.filter(event => event.agentId === expected))
  }

  /** Drop all ledger records. */
  clear(): void {
    this.ledger.length = 0
  }

  /** Markdown view matching the Python total and model-breakdown summary. */
  summary(): string {
    const totals = aggregate(this.ledger)
    const lines = [
      '# Cost Summary',
      '',
      'Total cost: $' + totals.costUsd.toFixed(4),
      'Total tokens: ' + formatInteger(totals.tokens)
        + ' (in: ' + formatInteger(totals.inputTokens)
        + ', out: ' + formatInteger(totals.outputTokens) + ')',
      'Events: ' + totals.turns,
      '',
    ]
    const models = this.byModel()
    if (Object.keys(models).length) {
      lines.push('## By Model')
      for (const [model, stats] of Object.entries(models).sort(([left], [right]) => left.localeCompare(right))) {
        lines.push('- **' + model + '**: $' + stats.costUsd.toFixed(4)
          + ' (' + stats.turns + ' turns, ' + formatInteger(stats.tokens) + ' tokens)')
      }
    }
    // Only ledgers that actually carry sources gain the section, so summaries
    // from callers that have not adopted the dimension stay byte-identical.
    const breakdown = this.sourceBreakdown()
    if (breakdown.housekeeping.turns || breakdown.main.turns) {
      lines.push('', '## By Source')
      for (const [source, stats] of Object.entries(breakdown.bySource)
        .sort(([left], [right]) => left.localeCompare(right))) {
        lines.push('- **' + source + '**: $' + stats.costUsd.toFixed(4)
          + ' (' + stats.turns + ' turns, ' + formatInteger(stats.tokens) + ' tokens)')
      }
      lines.push('Housekeeping share: ' + (breakdown.housekeepingCostShare * 100).toFixed(1) + '%')
    }
    return lines.join('\n')
  }

  /** Full JSON-safe ledger records, including cache and scope fields. */
  asRecords(): readonly CostEventRecord[] {
    return this.ledger.map(event => event.toRecord())
  }

  /** Python cost_tracker.py persistence records, omitting cache and scope extensions. */
  asDicts(): readonly LegacyCostEventRecord[] {
    return this.ledger.map(event => event.toLegacyRecord())
  }

  private append(event: CostEvent): CostEvent {
    this.ledger.push(event)
    return event
  }

  private nowTimestamp(): string {
    const now = this.clock()
    if (!(now instanceof Date) || Number.isNaN(now.valueOf())) {
      throw new RangeError('now must return a valid Date')
    }
    return now.toISOString()
  }

  private scopeOptions(options: RecordTurnOptions | RecordRawOptions): {
    readonly agentId?: string
    readonly sessionId?: string
    readonly source?: QuerySource
  } {
    const sessionId = scopeValue(options.sessionId, 'sessionId') ?? this.defaultSessionId
    const agentId = scopeValue(options.agentId, 'agentId') ?? this.defaultAgentId
    const source = sourceValue(options.source) ?? this.defaultSource
    return {
      ...(sessionId !== undefined ? { sessionId } : {}),
      ...(agentId !== undefined ? { agentId } : {}),
      ...(source !== undefined ? { source } : {}),
    }
  }
}

function aggregate(events: readonly CostEvent[]): CostAggregate {
  let inputTokens = 0
  let outputTokens = 0
  let cacheReadTokens = 0
  let cacheCreationTokens = 0
  let costUsd = 0
  for (const event of events) {
    inputTokens += event.inputTokens
    outputTokens += event.outputTokens
    cacheReadTokens += event.cacheReadTokens
    cacheCreationTokens += event.cacheCreationTokens
    costUsd += event.costUsd
  }
  return Object.freeze({
    turns: events.length,
    inputTokens,
    outputTokens,
    tokens: inputTokens + outputTokens,
    cacheReadTokens,
    cacheCreationTokens,
    cacheHitRate: cacheHitRateValues(inputTokens, cacheReadTokens),
    costUsd,
  })
}

function groupBy(
  events: readonly CostEvent[],
  keyFor: (event: CostEvent) => string,
): Readonly<Record<string, CostAggregate>> {
  const groups = new Map<string, CostEvent[]>()
  for (const event of events) {
    const key = keyFor(event)
    const values = groups.get(key) ?? []
    values.push(event)
    groups.set(key, values)
  }
  const result: Record<string, CostAggregate> = {}
  for (const [key, values] of groups) result[key] = aggregate(values)
  return Object.freeze(result)
}

function cacheHitRate(events: readonly CostEvent[]): number {
  const totals = aggregateTokenCounts(events)
  return cacheHitRateValues(totals.inputTokens, totals.cacheReadTokens)
}

function aggregateTokenCounts(events: readonly CostEvent[]): {
  readonly cacheReadTokens: number
  readonly inputTokens: number
} {
  let inputTokens = 0
  let cacheReadTokens = 0
  for (const event of events) {
    inputTokens += event.inputTokens
    cacheReadTokens += event.cacheReadTokens
  }
  return { inputTokens, cacheReadTokens }
}

function cacheHitRateValues(inputTokens: number, cacheReadTokens: number): number {
  const served = inputTokens + cacheReadTokens
  return served > 0 ? cacheReadTokens / served : 0
}

function finiteNumber(value: unknown, name: string): number {
  if (typeof value !== 'number' || !Number.isFinite(value)) {
    throw new RangeError(name + ' must be a finite number')
  }
  return value
}

function tokenCount(value: unknown, name: string): number {
  if (typeof value !== 'number' || !Number.isSafeInteger(value) || value < 0) {
    throw new RangeError(name + ' must be a non-negative safe integer')
  }
  return value
}

function stringValue(value: unknown, name: string): string {
  if (typeof value !== 'string') throw new TypeError(name + ' must be a string')
  return value
}

/**
 * Accept a source or its absence.
 *
 * The union is enforced by the compiler at every typed call site, so this only
 * has to reject the shapes a plain `string` cast could smuggle in — an empty
 * bucket key would silently merge tagged spend into a nameless group.
 */
function sourceValue(value: QuerySource | undefined): QuerySource | undefined {
  if (value === undefined) return undefined
  if (typeof value !== 'string' || !value.trim()) throw new TypeError('source must be a non-empty string')
  return value
}

function scopeValue(value: unknown, name: string): string | undefined {
  if (value === undefined) return undefined
  if (typeof value !== 'string' || !value.trim()) throw new TypeError(name + ' must be a non-empty string')
  return value
}

function timestampValue(value: unknown): string {
  if (typeof value !== 'string' || Number.isNaN(new Date(value).valueOf())) {
    throw new RangeError('timestamp must be a valid ISO timestamp')
  }
  return value
}

function formatInteger(value: number): string {
  return value.toLocaleString('en-US')
}
